### PR TITLE
Styleguidist & Button

### PR DESCRIPTION
**Pull request by lukeschunk**
# What?
This PR adds styleguidist to the component library which is awesome. It also adds some flow config and a new Button component (based on new designs). Also uses a generic theme (which will be added to) for the colors. 

## Change Type

* [x] Feature
* [ ] Chore
* [ ] Bug Fix

## Change Level

* [ ] major
* [ ] minor
* [ ] patch

## Further Information (screenshots, etc)
styleguidist_2

## Checklist

* [ ] Added tests / Task did not decrease code coverage
* [ ] If UI change, tested across supported browsers
* [ ] Documented manual test cases, if applicable

### CODE REVIEW

File: styleguide.config.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        exclude: /node_modules/,
+        loader: 'babel-loader',
+        options: babelRc,
+      },
+    ],
+  },
+}
+
+module.exports = {
+  components: 'src/**/[A-z]*.js',
```
Comment (lukeschunk) [inline styleguide.config.js]: couldn't get this regex quite right; I only want styleguidist to see certain files, not those with `index.js` and not those like `.styled.js` or `.types.js`

File: styleguide.config.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  //ignore: [
+  //'src/**/index.js',
+  //],
+  webpackConfig,
+  theme: {},
```
Comment (lukeschunk) [inline styleguide.config.js]: I started to play with this, but decided to leave that for a separate PR.

File: styleguide.config.js (hunk 1 above)
Comment (TomPridham) [reply styleguide.config.js]: you can prolly do `src/**/*.js|!src/**/index.js|!src/**/*.styled.js|!src/**/*.types.js`

File: .eslintrc.json [hunk 3]
```diff
@@ -1,12 +1,20 @@
 {
   "root": true,
   "plugins": [
-    "jane"
+    "jane",
+    "flowtype",
```
Comment (zacanger) [inline .eslintrc.json]: This should just be `plugins: [ jane ]`

File: .flowconfig [hunk 4]
```diff
@@ -0,0 +1,19 @@
+[ignore]
+.*/public/.*
+.*/lib/.*
+.*/build/.*
+.*/node_modules/styled-components/.*
+.*/.storybook/.*
```
Comment (zacanger) [inline .flowconfig]: These two lines can be removed

File: package.json [hunk 5]
```diff
@@ -33,11 +38,21 @@
     "jane-components",
     "react"
   ],
-  "dependencies": {},
+  "dependencies": {
+    "eslint-plugin-flowtype": "2.50.0",
```
Comment (zacanger) [inline package.json]: These two eslint plugins are deps of eslint-plugin-jane, so they can be removed. The two flow deps should be moved to devDeps.

File: package.json [hunk 6]
```diff
@@ -33,11 +38,21 @@
...
+    "eslint-plugin-flowtype-errors": "3.6.0",
+    "flow-bin": "0.77.0",
+    "flow-typed": "2.5.1"
+  },
   "devDependencies": {
     "babel-core": "6.26.3",
     "babel-jest": "23.4.0",
+    "babel-loader": "7.1.5",
+    "babel-plugin-styled-components": "1.5.1",
+    "babel-plugin-styled-name": "2.0.0",
```
Comment (zacanger) [inline package.json]: IIRC 2.x of this plugin is broken.

File: package.json [hunk 7]
```diff
@@ -54,14 +69,17 @@
...
     "react-dom": "16.4.1",
+    "react-styleguidist": "7.1.0",
     "rollup": "0.63.4",
     "rollup-plugin-babel": "3.0.7",
     "rollup-plugin-commonjs": "9.1.3",
     "rollup-plugin-node-resolve": "3.3.0",
     "rollup-plugin-peer-deps-external": "2.2.0",
-    "styled-components": "3.3.3"
+    "styled-components": "3.3.3",
+    "webpack": "4.16.2"
```
Comment (zacanger) [inline package.json]: If you're not using webpack to build an app, it shouldn't be here (it's a dep of styleguidist)

File: src/button/button.styled.js [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+type StyledProps = { ...ButtonProps, theme: Theme }
+
+export default styled.button`
+  white-space: nowrap;
+  background-color: ${({ theme, type }: StyledProps): string =>
+    type ? theme[type] : ''};
+  border-width: 0;
+  border-radius: 2px;
+  box-sizing: border-box;
+  color: ${({ theme }: StyledProps) => `${theme.white}`};
```
Comment (zacanger) [inline src/button/button.styled.js]: I'd really like it if we brought in a helper for this.

```javascript
const fromTheme = (key: string) => ({ theme }: { theme: typeof Theme }): string => theme[key]
```

File: src/theme.types.js [hunk 9]
```diff
@@ -0,0 +1,13 @@
+// @flow
+
+export type Theme = {
+  white: *,
```
Comment (zacanger) [inline src/theme.types.js]: Don't use `*`, it's deprecated. All of these are strings.

Comment (lukeschunk) [reply src/theme.types.js]: i agree these are all strings but I was getting a weird error (battled with it for at least 2 hours) and this is how I fixed it. 

The error: 
```
  20:27  error  Cannot create `ThemeProvider` element because string (see ./src/theme.types.js:9) is incompatible with mixed (see https://github.com/facebook/flow/blob/v0.77.0/lib/styled-components_v3.x.x.js#L210) in property `destructive` of property `theme`
```

File: src/button/button.styled.js (hunk 8 above)
Comment (TomPridham) [reply src/button/button.styled.js]: we should prolly just add that as a helper method we export as part of this lib

File: package.json (hunk 7 above)
Comment (lukeschunk) [reply package.json]: If I don't have it, I get: 

screen shot 2018-08-01 at 9 58 47 am

Comment (zacanger) [reply package.json]: Oh, I guess I'm wrong. That probably _should_ be a dependency of Styleguidist, but it looks like it's a peerDep.

File: src/theme.types.js (hunk 9 above)
Comment (TomPridham) [reply src/theme.types.js]: i just pushed a fix for this. I had to change the styled-components flow-typed file, so im gonna go make a pr to fix that at the source

Comment (TomPridham) [reply src/theme.types.js]: for more info check out the `invariant-by-default dictionaries` section 
https://flow.org/blog/2016/10/04/Property-Variance/

Comment (TomPridham) [reply src/theme.types.js]: someone beat me to it: https://github.com/flow-typed/flow-typed/pull/2556

File: flow-typed/npm/styled-components_v3.x.x.js [hunk 10]
```diff
@@ -207,7 +207,7 @@ type $npm$styledComponents$WithTheme = $npm$styledComponents$WithThemeReactCompo
   $npm$styledComponents$WithThemeReactComponentFunctionalUndefinedDefaultProps
 
 // ---- MISC ----
-type $npm$styledComponents$Theme = { [key: string]: mixed }
+type $npm$styledComponents$Theme = { +[key: string]: mixed }
```
Comment (zacanger) [inline flow-typed/npm/styled-components_v3.x.x.js]: Don't ever edit files under `flow-typed` by hand. Treat them like node_modules: if there's a bug, fix it upstream.

File: .flowconfig [hunk 11]
```diff
@@ -3,6 +3,7 @@
 .*/lib/.*
 .*/build/.*
 .*/node_modules/styled-components/.*
+.*/node_modules/findup/.*
 .*/.storybook/.*
```
Comment (zacanger) [inline .flowconfig]: These storybook lines can still be removed.

File: flow-typed/npm/styled-components_v3.x.x.js (hunk 10 above)
Comment (lukeschunk) [reply flow-typed/npm/styled-components_v3.x.x.js]: hmm weird i didn't edit this by hand...

File: .flowconfig (hunk 11 above)
Comment (lukeschunk) [reply .flowconfig]: lol i hadn't pushed up my changes

File: src/button/button.styled.js (hunk 8 above)
Comment (lukeschunk) [reply src/button/button.styled.js]: i'm pulling it in, but I think for situations like this I won't be able to use it, correct? 
```
    background-color: ${({ theme, type }: StyledProps) =>
      type ? theme[type + 'Darker'] : ''};
```

File: flow-typed/npm/styled-components_v3.x.x.js (hunk 10 above)
Comment (TomPridham) [reply flow-typed/npm/styled-components_v3.x.x.js]: i did this. there is a pr to fix it. this fixes the * type errors in theme for now

Comment (TomPridham) [reply flow-typed/npm/styled-components_v3.x.x.js]: a pr in flow-typed

File: src/button/button.styled.js (hunk 8 above)
Comment (zacanger) [reply src/button/button.styled.js]: Correct

File: src/button/button.js [hunk 12]
```diff
@@ -0,0 +1,26 @@
+// @flow
+
+import * as React from 'react'
+import { ThemeProvider } from 'styled-components'
+import theme from '../theme'
+import StandardButton from './button.styled'
+
+export type ButtonProps = {
+  disabled?: boolean,
+  style?: *,
```
Comment (TomPridham) [inline src/button/button.js]: i can't remember if i told you to do this, but i saw @whiteinge doing `{}` in a couple places and looked it up and it's the better option. https://flow.org/en/docs/types/objects/#toc-object-type
